File: apps/web/assets/src/lib/solana-tx.js
// Solana legacy transaction builder (browser-only, no @solana/web3.js dep).
//
// Wire format (legacy, message v0):
//   tx       = signatures || message
//   signatures = compact-u16 N || N × 64-byte Ed25519 sigs (msg-order)
//   message  = header(3) || accountKeys || recentBlockhash(32) || instructions
//     header = numRequiredSigs(u8) || numReadonlySigned(u8) || numReadonlyUnsigned(u8)
//     accountKeys  = compact-u16 N || N × 32-byte pubkeys
//     instructions = compact-u16 N || N × { programIdIndex(u8) || accounts(compact+N×u8) || data(compact+N×u8) }
//
// Compact-u16 (also called short-vec): 1–3 bytes, 7 bits per byte, MSB
// continuation. Used both for top-level counts and per-instruction lens.
//
// AccountKeys ordering rules (canonical):
//   1) Writable signers
//   2) Read-only signers
//   3) Writable non-signers
//   4) Read-only non-signers (program IDs always end up here)
// Payer is forced to index 0 (first writable signer).
(function () {
  function encodeCompactU16(n) {
    if (n < 0 || n > 0xffff) throw new Error("compact-u16: out of range " + n);
    const out = [];
    let rem = n;
    for (let i = 0; i < 3; i++) {
      const lo = rem & 0x7f;
      rem >>= 7;
      if (rem === 0) { out.push(lo); break; }
      out.push(lo | 0x80);
    }
    return new Uint8Array(out);
  }

  function decodeCompactU16(buf, off) {
    let n = 0, shift = 0, consumed = 0;
    for (let i = 0; i < 3; i++) {
      const b = buf[off + i];
      consumed++;
      n |= (b & 0x7f) << shift;
      if ((b & 0x80) === 0) break;
      shift += 7;
    }
    return { value: n, consumed };
  }

  function eq32(a, b) {
    if (a.length !== 32 || b.length !== 32) return false;
    for (let i = 0; i < 32; i++) if (a[i] !== b[i]) return false;
    return true;
  }
  function findKey(keys, pk) {
    for (let i = 0; i < keys.length; i++) if (eq32(keys[i], pk)) return i;
    return -1;
  }

  function classifyAccounts(payer, instructions) {
    // Track each unique pubkey with the strongest (signer, writable) it appears as.
    const meta = new Map(); // key = hex(pubkey) → { pk, isSigner, isWritable, isProgram }
    const hexOf = (pk) => Array.from(pk, (b) => b.toString(16).padStart(2, "0")).join("");

    // Payer is always signer + writable + first.
    meta.set(hexOf(payer), { pk: payer, isSigner: true, isWritable: true, isProgram: false });

    for (const ix of instructions) {
      // Program ID: read-only non-signer.
      const pk = ix.programId;
      const k = hexOf(pk);
      if (!meta.has(k)) meta.set(k, { pk, isSigner: false, isWritable: false, isProgram: true });
      // Note: a program could theoretically be referenced as an instruction account too,
      // but we keep its isProgram flag (it sorts to read-only-non-signer regardless).

      for (const a of ix.accounts) {
        const hk = hexOf(a.pubkey);
        const prior = meta.get(hk);
        if (!prior) {
          meta.set(hk, { pk: a.pubkey, isSigner: !!a.isSigner, isWritable: !!a.isWritable, isProgram: false });
        } else {
          // Upgrade flags if any role demands it.
          prior.isSigner   = prior.isSigner   || !!a.isSigner;
          prior.isWritable = prior.isWritable || !!a.isWritable;
        }
      }
    }

    // Sort: payer first, then by (signer desc, writable desc, isProgram asc).
    const payerHex = hexOf(payer);
    const all = Array.from(meta.values());
    all.sort((a, b) => {
      // payer always wins
      if (hexOf(a.pk) === payerHex) return -1;
      if (hexOf(b.pk) === payerHex) return  1;
      const sa = a.isSigner ? 0 : 1, sb = b.isSigner ? 0 : 1;
      if (sa !== sb) return sa - sb;
      const wa = a.isWritable ? 0 : 1, wb = b.isWritable ? 0 : 1;
      if (wa !== wb) return wa - wb;
      const pa = a.isProgram ? 1 : 0, pb = b.isProgram ? 1 : 0;
      return pa - pb;
    });

    let numRequiredSigs = 0, numReadonlySigned = 0, numReadonlyUnsigned = 0;
    for (const m of all) {
      if (m.isSigner) {
        numRequiredSigs++;
        if (!m.isWritable) numReadonlySigned++;
      } else if (!m.isWritable) {
        numReadonlyUnsigned++;
      }
    }
    return {
      header: { numRequiredSigs, numReadonlySigned, numReadonlyUnsigned },
      keys: all.map((m) => m.pk),
    };
  }

  function concat(parts) {
    let total = 0;
    for (const p of parts) total += p.length;
    const out = new Uint8Array(total);
    let off = 0;
    for (const p of parts) { out.set(p, off); off += p.length; }
    return out;
  }

  class LegacyTx {
    constructor({ header, keys, recentBlockhash, instructions }) {
      this.header = header;
      this.keys = keys;
      this.recentBlockhash = recentBlockhash;
      this.instructions = instructions;
      this._cachedMsg = null;
    }
    serializeMessage() {
      if (this._cachedMsg) return this._cachedMsg;
      const parts = [];
      parts.push(new Uint8Array([this.header.numRequiredSigs, this.header.numReadonlySigned, this.header.numReadonlyUnsigned]));
      parts.push(encodeCompactU16(this.keys.length));
      for (const k of this.keys) parts.push(k);
      parts.push(this.recentBlockhash);
      parts.push(encodeCompactU16(this.instructions.length));
      for (const ix of this.instructions) {
        parts.push(new Uint8Array([ix.programIdIndex]));
        parts.push(encodeCompactU16(ix.accountIndexes.length));
        parts.push(new Uint8Array(ix.accountIndexes));
        parts.push(encodeCompactU16(ix.data.length));
        parts.push(ix.data);
      }
      this._cachedMsg = concat(parts);
      return this._cachedMsg;
    }
    async sign(signers) {
      const msg = this.serializeMessage();
      const numRequired = this.header.numRequiredSigs;
      const sigs = new Array(numRequired).fill(null);
      for (const s of signers) {
        const idx = findKey(this.keys, s.pubkey);
        if (idx < 0)            throw new Error("signer not in accountKeys");
        if (idx >= numRequired) throw new Error("signer at position " + idx + " is not a required signer");
        sigs[idx] = await s.sign(msg);
      }
      for (let i = 0; i < numRequired; i++) {
        if (!sigs[i]) {
          // Pad with zeros — caller can chain another sign() later if needed.
          sigs[i] = new Uint8Array(64);
        }
      }
      const parts = [encodeCompactU16(numRequired)];
      for (const s of sigs) parts.push(s instanceof Uint8Array ? s : new Uint8Array(s));
      parts.push(msg);
      return concat(parts);
    }
  }

  function buildLegacyTx({ payer, recentBlockhash, instructions }) {
    if (!(payer instanceof Uint8Array) || payer.length !== 32) {
      throw new Error("payer must be 32-byte Uint8Array");
    }
    if (!(recentBlockhash instanceof Uint8Array) || recentBlockhash.length !== 32) {
      throw new Error("recentBlockhash must be 32 bytes");
    }
    if (!Array.isArray(instructions) || instructions.length === 0) {
      throw new Error("at least one instruction required");
    }

    const { header, keys } = classifyAccounts(payer, instructions);

    const compiled = instructions.map((ix) => {
      const programIdIndex = findKey(keys, ix.programId);
      if (programIdIndex < 0) throw new Error("programId missing from accountKeys (bug)");
      const accountIndexes = ix.accounts.map((a) => {
        const i = findKey(keys, a.pubkey);
        if (i < 0) throw new Error("instruction account missing from keys (bug)");
        return i;
      });
      return {
        programIdIndex,
        accountIndexes,
        data: ix.data instanceof Uint8Array ? ix.data : new Uint8Array(ix.data),
      };
    });

    return new LegacyTx({ header, keys, recentBlockhash, instructions: compiled });
  }

  const _SolanaTx = {
    buildLegacyTx,
    encodeCompactU16,
    decodeCompactU16,
    classifyAccounts,
    LegacyTx,
  };
  if (typeof globalThis !== "undefined") globalThis.SolanaTx = _SolanaTx;
})();

File: core/onchain/solana_tx.zig
//! Solana legacy transaction builder.
//!
//! Mirrors solana-tx.js (apps/web/assets/src/lib/solana-tx.js).
//!
//! Wire format (legacy tx):
//!   tx       = signatures || message
//!   signatures = compact-u16 N || N × 64-byte Ed25519 sigs
//!   message  = header(3) || accountKeys(compact+N×32) ||
//!              recentBlockhash(32) || instructions
//!     header = numRequiredSigs(u8) || numReadonlySigned(u8) || numReadonlyUnsigned(u8)
//!     instructions = compact-u16 N || N × {
//!       programIdIndex(u8) || accounts(compact+N×u8) || data(compact+N×u8)
//!     }
//!
//! Compact-u16 (short-vec): 1–3 bytes, 7 bits per byte, MSB = continuation.

const std = @import("std");
const crypto_mod = @import("../security/crypto.zig");
const types = @import("../protocol/types.zig");

pub const Pubkey = [32]u8;
pub const Blockhash = [32]u8;

pub const AccountMeta = struct {
    pubkey: Pubkey,
    is_signer: bool,
    is_writable: bool,
};

pub const Instruction = struct {
    program_id: Pubkey,
    accounts: []const AccountMeta,
    data: []const u8,
};

pub const Header = struct {
    num_required_sigs: u8,
    num_readonly_signed: u8,
    num_readonly_unsigned: u8,
};

pub const CompiledInstruction = struct {
    program_id_index: u8,
    account_indexes: []const u8,
    data: []const u8,
};

// ── Compact-U16 (short-vec) ───────────────────────────────────────────────

pub fn writeCompactU16(writer: anytype, value: usize) !void {
    var val: usize = value;
    while (true) {
        var byte: u8 = @intCast(val & 0x7F);
        val >>= 7;
        if (val > 0) byte |= 0x80;
        try writer.writeByte(byte);
        if (val == 0) break;
    }
}

pub fn decodeCompactU16(buf: []const u8, off: usize) struct { value: usize, consumed: usize } {
    var n: usize = 0;
    var shift: u6 = 0;
    var consumed: usize = 0;
    var i: usize = 0;
    while (i < 3) : (i += 1) {
        const b = buf[off + i];
        consumed += 1;
        n |= (@as(usize, b & 0x7F)) << shift;
        if ((b & 0x80) == 0) break;
        shift += 7;
    }
    return .{ .value = n, .consumed = consumed };
}

// ── Account classification ────────────────────────────────────────────────

const AccountEntry = struct {
    pubkey: Pubkey,
    is_signer: bool,
    is_writable: bool,
};

fn pubkeyEq(a: *const Pubkey, b: *const Pubkey) bool {
    return std.mem.eql(u8, a, b);
}

/// Classify and sort accounts for a Solana legacy tx.
/// Order: writable-signers → readonly-signers → writable-nonsigners → readonly-nonsigners.
/// Payer is forced to index 0.
pub fn classifyAccounts(
    allocator: std.mem.Allocator,
    payer: *const Pubkey,
    instructions: []const Instruction,
) !struct { header: Header, keys: []Pubkey } {
    var entries = std.ArrayListUnmanaged(AccountEntry){};
    defer entries.deinit(allocator);

    // Helper: find or insert an entry, upgrading flags.
    const findOrInsert = struct {
        fn call(list: *std.ArrayListUnmanaged(AccountEntry), alloc: std.mem.Allocator, pk: *const Pubkey, signer: bool, writable: bool) !void {
            for (list.items) |*e| {
                if (pubkeyEq(&e.pubkey, pk)) {
                    if (signer) e.is_signer = true;
                    if (writable) e.is_writable = true;
                    return;
                }
            }
            try list.append(alloc, .{ .pubkey = pk.*, .is_signer = signer, .is_writable = writable });
        }
    }.call;

    // Payer: always writable signer, always index 0.
    try findOrInsert(&entries, allocator, payer, true, true);

    for (instructions) |ix| {
        // Program ID: readonly non-signer.
        try findOrInsert(&entries, allocator, &ix.program_id, false, false);
        // Instruction accounts.
        for (ix.accounts) |a| {
            try findOrInsert(&entries, allocator, &a.pubkey, a.is_signer, a.is_writable);
        }
    }

    // Sort: payer first, then (signer desc, writable desc).
    std.mem.sort(AccountEntry, entries.items, payer, struct {
        fn lessThan(pay: *const Pubkey, a: AccountEntry, b: AccountEntry) bool {
            const a_is_payer = pubkeyEq(&a.pubkey, pay);
            const b_is_payer = pubkeyEq(&b.pubkey, pay);
            if (a_is_payer != b_is_payer) return a_is_payer;
            // Signer > non-signer
            if (a.is_signer != b.is_signer) return a.is_signer;
            // Writable > readonly
            if (a.is_writable != b.is_writable) return a.is_writable;
            return false;
        }
    }.lessThan);

    // Compute header counters.
    var num_required_sigs: u8 = 0;
    var num_readonly_signed: u8 = 0;
    var num_readonly_unsigned: u8 = 0;
    for (entries.items) |e| {
        if (e.is_signer) {
            num_required_sigs += 1;
            if (!e.is_writable) num_readonly_signed += 1;
        } else if (!e.is_writable) {
            num_readonly_unsigned += 1;
        }
    }

    // Build final key list.
    const keys = try allocator.alloc(Pubkey, entries.items.len);
    for (entries.items, 0..) |e, i| keys[i] = e.pubkey;

    return .{
        .header = .{
            .num_required_sigs = num_required_sigs,
            .num_readonly_signed = num_readonly_signed,
            .num_readonly_unsigned = num_readonly_unsigned,
        },
        .keys = keys,
    };
}

fn findKey(keys: []const Pubkey, pk: *const Pubkey) ?u8 {
    for (keys, 0..) |k, i| {
        if (pubkeyEq(&k, pk)) return @intCast(i);
    }
    return null;
}

// ── Transaction builder ───────────────────────────────────────────────────

/// Build a serialized Solana legacy transaction (unsigned — signature bytes zeroed).
/// Returns owned byte slice. Caller should call signTx() afterwards.
pub fn buildLegacyTx(
    allocator: std.mem.Allocator,
    payer: *const Pubkey,
    blockhash: *const Blockhash,
    instructions: []const Instruction,
) ![]u8 {
    const classified = try classifyAccounts(allocator, payer, instructions);
    defer allocator.free(classified.keys);

    const keys = classified.keys;
    const header = classified.header;
    const num_sigs = header.num_required_sigs;

    var buf = std.ArrayListUnmanaged(u8){};
    errdefer buf.deinit(allocator);
    const writer = buf.writer(allocator);

    // Signatures placeholder.
    try writeCompactU16(writer, num_sigs);
    try buf.appendNTimes(allocator, 0, @as(usize, num_sigs) * 64);

    // Message header.
    try writer.writeByte(header.num_required_sigs);
    try writer.writeByte(header.num_readonly_signed);
    try writer.writeByte(header.num_readonly_unsigned);

    // Account keys.
    try writeCompactU16(writer, keys.len);
    for (keys) |k| try writer.writeAll(&k);

    // Recent blockhash.
    try writer.writeAll(blockhash);

    // Instructions.
    try writeCompactU16(writer, instructions.len);
    for (instructions) |ix| {
        const prog_idx = findKey(keys, &ix.program_id) orelse return error.ProgramNotInKeys;
        try writer.writeByte(prog_idx);

        try writeCompactU16(writer, ix.accounts.len);
        for (ix.accounts) |a| {
            const acc_idx = findKey(keys, &a.pubkey) orelse return error.AccountNotInKeys;
            try writer.writeByte(acc_idx);
        }

        try writeCompactU16(writer, ix.data.len);
        try writer.writeAll(ix.data);
    }

    return buf.toOwnedSlice(allocator);
}

/// Sign a serialized legacy tx in-place. The message starts at byte 65
/// (1 byte compact-u16 for sig count + 64 bytes for the sig placeholder).
/// Works for single-signer txs (the common case for CLI commands).
pub fn signTx(tx_buf: []u8, keypair: *const types.Keypair) void {
    // Message: everything after the 1-byte compact-u16 sig count (=0x01) and 64-byte sig slot.
    const message = tx_buf[65..];
    const sig = crypto_mod.sign(message, keypair);
    @memcpy(tx_buf[1..65], &sig);
}

// ── Tests ────────────────────────────────────────────────────────────────

test "encodeCompactU16" {
    const allocator = std.testing.allocator;

    var buf = std.ArrayListUnmanaged(u8){};
    defer buf.deinit(allocator);

    // 0 → 0x00
    try writeCompactU16(buf.writer(allocator), 0);
    try std.testing.expectEqualSlices(u8, &[_]u8{0x00}, buf.items);
    buf.clearRetainingCapacity();

    // 127 → 0x7F
    try writeCompactU16(buf.writer(allocator), 127);
    try std.testing.expectEqualSlices(u8, &[_]u8{0x7F}, buf.items);
    buf.clearRetainingCapacity();

    // 128 → 0x80 0x01
    try writeCompactU16(buf.writer(allocator), 128);
    try std.testing.expectEqualSlices(u8, &[_]u8{0x80, 0x01}, buf.items);
    buf.clearRetainingCapacity();

    // 125 (instruction data length in fixture) → single byte 0x7D
    try writeCompactU16(buf.writer(allocator), 125);
    try std.testing.expectEqualSlices(u8, &[_]u8{0x7D}, buf.items);
}

test "buildLegacyTx: single instruction, single signer" {
    const allocator = std.testing.allocator;

    var payer: Pubkey = undefined;
    std.crypto.random.bytes(&payer);
    var prog: Pubkey = undefined;
    std.crypto.random.bytes(&prog);
    var bh: Blockhash = undefined;
    std.crypto.random.bytes(&bh);

    const data = [_]u8{ 1, 2, 3 };
    const ix = Instruction{
        .program_id = prog,
        .accounts = &.{},
        .data = &data,
    };

    const tx = try buildLegacyTx(allocator, &payer, &bh, &[_]Instruction{ix});
    defer allocator.free(tx);

    // Must start with compact-u16(1) = 0x01 then 64 zero bytes.
    try std.testing.expectEqual(@as(u8, 0x01), tx[0]);
    var i: usize = 1;
    while (i < 65) : (i += 1) try std.testing.expectEqual(@as(u8, 0), tx[i]);
}
